### DISCUSSION TITLE
SubSonic plugin for Upmpdcli.

### DISCUSSION THREAD

**Discussion by GioF71** [navidrome/navidrome#2324] [Show and tell]
Hello from a happy Navidrome user.

With some good support from the creator of upmpdcli, I have created a plugin for upmpdcli which enables the use of Navidrome (generally and hopefully all other servers compliant with the subsonic api) with an upnp/dlna renderer.

Some screenshots: [Upplay on Linux](https://github.com/GioF71/upmpdcli-docker/wiki/Upplay-screenshots) and [Linn Kazoo on Windows](https://github.com/GioF71/upmpdcli-docker/wiki/Linn-Kazoo-Screenshots)

References to upmpdcli and mpd:

- Upmpdcli [home page](https://www.lesbonscomptes.com/upmpdcli/)
- SubSonic Plugin [config](https://www.lesbonscomptes.com/upmpdcli/pages/upmpdcli-manual.html#UPMPDCLI-MS-SUBSONIC) for [upmpdcli](https://www.lesbonscomptes.com/upmpdcli/)
- Upmpdcli repos on Framagit: [upstream](https://framagit.org/medoc92/upmpdcli) & [my fork](https://framagit.org/GioF71/upmpdcli)
- My upmpdcli docker image: [GitHub Repo](https://github.com/GioF71/upmpdcli-docker), [Docker Hub](https://hub.docker.com/r/giof71/upmpdcli)
- My mpd docker image: [GitHub Repo](https://github.com/GioF71/mpd-alsa-docker), [Docker Hub](https://hub.docker.com/r/giof71/mpd-alsa)
- [Music Player Daemon](https://musicpd.org/)

The plugin is based on my library [subsonic-connector](https://github.com/GioF71/subsonic-connector) which in turn is built around [py-sonic](https://github.com/crustymonkey/py-sonic).  

You will need a running mpd/upmpdcli (in renderer mode) as well of course, or possibly any other upnp/dlna renderer.  
Kodi as a renderer works.

Sample docker compose, configured for the navidrome demo:

```text
---
version: "3"

services:
  upmpdcli:
    image: giof71/upmpdcli:latest
    container_name: upmpdcli-navidrome-demo
    network_mode: host
    environment:
      - PUID=1000 #optional
      - PGID=1000 #optional
      - PORT_OFFSET=4 #optional
      - UPNPAV=0 #not a renderer
      - OPENHOME=0 #not a renderer
      - FRIENDLY_NAME=NavidromeDemo
      - SUBSONIC_ENABLE=yes
      - SUBSONIC_AUTOSTART=yes
      - SUBSONIC_BASE_URL=${SUBSONIC_BASE_URL}
      - SUBSONIC_PORT=${SUBSONIC_PORT}
      - SUBSONIC_USER=${SUBSONIC_USER}
      - SUBSONIC_PASSWORD=${SUBSONIC_PASSWORD}
      - SUBSONIC_ITEMS_PER_PAGE=28 #optional
      - SUBSONIC_APPEND_YEAR_TO_ALBUM=Yes #optional
      - SUBSONIC_APPEND_CODECS_TO_ALBUM=Yes #optional
      - SUBSONIC_WHITELIST_CODECS=alac,wav,flac,dsf,ape #optional
      - SUBSONIC_DOWNLOAD_PLUGIN=yes #optional, will download updated plugin from the specified branch
      - SUBSONIC_PLUGIN_BRANCH=latest #optional, it is the branch to be downloaded if SUBSONIC_DOWNLOAD_PLUGIN is set to yes
    restart: unless-stopped
```

Create the .env file:

```text
SUBSONIC_BASE_URL=https://demo.navidrome.org
SUBSONIC_PORT=443
SUBSONIC_USER=demo
SUBSONIC_PASSWORD=demo
```

This server will be somehow slow, but if you run the container against an instance on your network, it will be a lot faster.
I will create a sample video if this can be useful.

For better integration, you might want to also run the scrobbler (a scrobbler to the Subsonic server). See [this](https://github.com/navidrome/navidrome/discussions/2324#discussioncomment-5886980) post from this discussion.

Enjoy!

Edit: typos, removed reference to my homelab docker registry, added relevant links, add screenshots

**Comment by andrewdunndev** [comment]
@GioF71 I'd be interested in the sample video!

**Reply by GioF71** [reply]
I'll see if I can prepare it in the next few days!

**Reply by GioF71** [reply]
Sorry for not having prepared the video yet. I thought it would have been easier to have a few hours of consecutive spare time.
The video would be a screencast anyway with voiceover (possibly).
Anyway, did you have a chance to try the solution?
Let me know if I can help you somehow.

**Reply by andrewdunndev** [reply]
No problem, I appreciate what you've done so far and have not had a chance to try it yet. The reason I'd wanted to see a video was trying to understand is how the control flow works, I've not used uPnP with MPD before. I'm aware of some players (e.g. jplay) that can submit tracks from sources (e.g. qobuz) to mpd via uPnP.

with the upmpdcli method do you point mpd at the navidrome instance and mpd indexes content, then control it via some mpd controller? Or do you need a uPnP based player that submits tracks to an mpd instance?

**Reply by GioF71** [reply]
> No problem, I appreciate what you've done so far and have not had a chance to try it yet. The reason I'd wanted to see a video was trying to understand is how the control flow works, I've not used uPnP with MPD before. I'm aware of some players (e.g. jplay) that can submit tracks from sources (e.g. qobuz) to mpd via uPnP.
> 
> with the upmpdcli method do you point mpd at the navidrome instance and mpd indexes content, then control it via some mpd controller? Or do you need a uPnP based player that submits tracks to an mpd instance?

I know, I am sorry, I can't promise anything but I am thinking about this every day.
About your question, no, mpd does not need to know anything about Navidrome. There are no indexes to build or keep updated other than Navidrome itself.
And MPD with upmpdcli IS the upnp/dlna player, you don't need another player. Upmpdcli, in "renderer" mode, is the "bridge" between UPnP/DLNA media servers and mpd. But upmpdcli can also be a UPnP/DLNA server, which is what you obtain by running the compose file in the first post.
If you run that compose file (changing the .env file with the correct setting for your navidrome server), you get a UPnP/DLNA server on your network.
MPD+Upmpdcli is one way to create a UPnP/DLNA renderer. This solution is used by Volumio, Moode Audio, and other commercial devices, notably Aurender. Some still use mpd illegally, see [here](https://www.musicpd.org/commercial.html).
Other renderers might be even commercial devices, like the Wiim Mini/Pro, some Cambridge Audio network streamers, just to name a few. Not necessarily those devices are built around mpd of course.
A Google Chromecast (Audio or regular) can also be used as a DLNA renderer, but for this you would need an instance of BubbleUpnpServer (free), or directly if you have BubbleUpnp on a smartphone.
And, maybe even your smart tv can be used as a renderer if you use BubbleUpnp on your smartphone. I just tried it :-)

I like my mpd instances to always have upmpdcli on top, so they are free of the burden of handling/updating the library, which might be a heavy task for low powered devices like the PIs, especially if the library is big enough.
Also, if your renderers handle the library, and the content has changed while the renderers were off, you would probably need to update the library before being able to play any new content. This is something that you can avoid with mpd/upmpdcli used as a pure player: the player can immediately play the latest stuff without have to reindex upfront.

Don't get me wrong, mpd is also great on its own. I use it on my desktop with mympd. This computer (nothing special) is fast enough so that the task of updating the library is painless. Still, I use upmpdcli also on the desktop, I find it extremely convenient. 

Last thing, you can access Qobuz with upmpdcli, even in hi-res.

**Reply by andrewdunndev** [reply]
Thanks for the detailed reply!

What controller/player are you using that can be aware of the MPD+Upmpdcli renderer? Like, I can't set this as a playback in the standard navidrome web interface right?

**Reply by GioF71** [reply]
No, and I believe that's something out of scope for Navidrome. Maybe direct mpd support could be considered, for a JukeBox feature, and it would be very nice. It is 'possible', I created a mock app with Django that appends url to the mpd playlist and triggers playback and it works. But I do not mean to say that it would be easy to add the feature in Navidrome of course.
We should ask the developer, but I believe that JukeBox is a feature he is thinking about from a few tickets I saw. But I am not really updated with the status of these requests.

About the controller applications, on Linux you might use [upplay](https://www.lesbonscomptes.com/upplay/pages/downloads.html) which is available on ubuntu, debian and arch (+ derived).
On windows, you can use upplay as well, and another that works well for me is Linn Kazoo (in the windows store iirc), even if it only works with av renderers.
On android, you can use BubbleUpnp (best IMO), mConnect (only works against av renderers, not openhome). HiFi Cast also seems to work. There are many others, unfortunately I don't use them all. BubbleUpnp is a good choice, even the premium version costs a few bucks and it's totally worth to buy even only to avoid any ad and limitation.
On apple devices, I can't help you, but from a quick search it seems that mConnect is also available, so it's probably worth checking.

p.s.: about openhome and av renderers, I wanted to specify that upmpdcli can create both type of renderers.

**Reply by GioF71** [reply]
> @GioF71 I'd be interested in the sample video!

Finally here. See [this post](https://github.com/navidrome/navidrome/discussions/2324#discussioncomment-5898220).

**Comment by deluan** [comment]
Great work! Wanna provide a PR to the list of apps/clients in the documentation? https://www.navidrome.org/docs/overview/#apps

**Reply by GioF71** [reply]
Sure! Thank you!

**Comment by GioF71** [comment]
Hello, happy to announce that with the latest release (which is pulled using the suggested compose file), it is possible to browse and play directly from kodi.
Kodi unfortunately has a quirky behavior: it caches the responses from the upnp server (I can't really imagine why), so some lists won't change if the content changes while kodi is running.
Also, it won't work with the upnp server if this is restarted while kodi is running.
See [this](https://forum.kodi.tv/showthread.php?tid=330398) page for more details. This issue has been reported years ago.
A possible workaround, other than restarting kodi, is to disable and enable again the upnp/dlna support. This will reset the cache.

I consider using kodi via upnp for SubSonic quite a marginal use-case, but anyway at least now it kinda works.
Kodi already works perfectly with the subsonic plugin when used as a dlna renderer (you will need a control point like upplay, bubbleupnp, kazoo, ...).
A better (IMO) option for this scenario is the kodi Subsonic plugin available [here](https://github.com/warwickh/plugin.audio.subsonic).

Edit: minor changes, typos

**Reply by andrewdunndev** [reply]
Something nice about this is getting playback on television type devices.

I'd love to be able to use navidrome directly from appletv.

**Reply by GioF71** [reply]
> Something nice about this is getting playback on television type devices.
> 
> I'd love to be able to use navidrome directly from appletv.

I don't have an Apple TV, but maybe [this](https://apps.apple.com/bt/app/video-tv-cast-dlna-upnp-hd/id978357787) app will allow to play from upnp/sources.

**Comment by GioF71** [comment]
Hello, version 0.1.4 is available with a few improvements:

- kodi should keep working across upmpdcli restarts, because the generated entry ids are not hashcodes like before
- on startup, the total number of loaded albums (in steps of 500, max allowed by the subsonic api) is displayed on console output
- autostart is functional now
- general code cleanup

This version is download with the suggested configuration, through the variables:

- SUBSONIC_DOWNLOAD_PLUGIN=yes
- SUBSONIC_PLUGIN_BRANCH=latest

**Comment by GioF71** [comment]
Hello, version 0.1.5 is available with a few improvements:

- Support for Playlists
- Better selection of art for genres

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

This version requires the container image version 2023-05-09 and on, since it needs subsonic-connector version 0.1.11, so remember to pull the newest image, or the new "Playlists" entry will not work.

**Comment by GioF71** [comment]
Hello, version 0.1.6 is available with a few improvements:

- Support for Recently Played
- Better selection of art for genres

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to always update the docker image.

**Comment by GioF71** [comment]
Hello, we have a Subsonic scrobbler for MPD. It will help you keep track of what you play with the subsonic plugin for a better integration with the functionalities of Navidrome (and potentially other Subsonic servers).

See the [repository](https://github.com/GioF71/mpd-subsonic-scrobbler), and the [issue](https://github.com/navidrome/navidrome/issues/2347) that lead to this project.

Docker is the preferred way of running the application. However, it should be fairly simple to run the application directly in case you want to do that.

Just as an example, I report here a sample configuration, which requires a `.env` file with the subsonic parameters. Refer to the [repository](https://github.com/GioF71/mpd-subsonic-scrobbler) for more details:

```text
---
version: "3"

networks:
  mpd:
    external: true

services:
  scrobbler:
    image: giof71/mpd-subsonic-scrobbler:latest
    container_name: subsonic-scrobbler-d10
    networks:
      - mpd
    environment:
      - PYTHONUNBUFFERED=1
      - MPD_HOST=mpd-d10
      - MPD_PORT=6600
      - _SUBSONIC_BASE_URL=${SUBSONIC_BASE_URL}
      - _SUBSONIC_PORT=${SUBSONIC_PORT}
      - _SUBSONIC_USER=${SUBSONIC_USER}
      - _SUBSONIC_PASSWORD=${SUBSONIC_PASSWORD}
      - VERBOSE=0
    restart: unless-stopped
```

**Notice**: the docker hub repo was created with a different name, now I have created a new one matching the project name. The correct image name is `giof71/mpd-subsonic-scrobbler`.
Please check your docker commands. Sorry for the inconvenience.

**Comment by GioF71** [comment]
Hello, version 0.1.7 is available with a few improvements:

- New view: Highest Rated
- New view: Favourites
- New view: Most Played

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Finally here's the video. Took an effort to me, I am really not used to do this stuff.
Forgive the Daffy Duck voice. Anyway I hope it helps. I will add more contents with BubbleUpnp in the next few days, but maybe in form of screenshots.

Link to the [video](https://youtu.be/tIhK3ub9h-0).

**Comment by GioF71** [comment]
Hello, version 0.1.8 is available with a few improvements:

- Resiliency (if subsonic server goes down, the media server will recover without needing a restart)
- Faster caching
- General cleanup of unused code

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.1.9 is available with a few improvements:

- Support for Internet Radios (requires subsonic-connector version 0.1.14 so pulling the new upmpdcli image is mandatory)

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.1.10 is available with a few improvements:

- Support for Random Songs. For a clicked random song, you will be presented with the song, the album, and the artist. 
- Some type of Internet Radios might now play with Kodi

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.1.11 is available with a few improvements:

- Bug: Loop in Genre/Artist/Albums view if more than 500 albums per genre (offset was updated but not used in the query)
- Feature: Allowing "D" as prefix for album split in discs (Report Real Path must be enabled in order to distinguish albums with the same tags, but different folders)

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.1.12 is available with a few improvements:

- Support for top songs
- Support for similar artists

These features require subsonic-connector version 0.1.15 so pulling the new upmpdcli image is mandatory.
These features also require last.fm integration in Navidrome.

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.1.13 is available with a few improvements:

- Albums per Artist is now paginated and faster to load

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.1.14 is available with a few improvements:

- Add artist and album radio, in both navigable and flat mode
- Random songs also in "list mode" so we can directly see the songs
- Small bugs fixed

These features require subsonic-connector version 0.1.16 so pulling the new upmpdcli image is mandatory.
This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, bugfix version 0.1.15 is available. As it was very quick I am reporting changes from 0.1.14

- From 0.1.14: Add artist and album radio, in both navigable and flat mode
- From 0.1.14: Random songs also in "list mode" so we can directly see the songs
- From 0.1.14: Small bugs fixed
- BugFix Search was broken on 0.1.14, fixed now.

These features require subsonic-connector version 0.1.16 so pulling the new upmpdcli image is mandatory.
This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

We have screenshots for [Upplay on Linux](https://github.com/GioF71/upmpdcli-docker/wiki/Upplay-screenshots) and [Linn Kazoo on Windows](https://github.com/GioF71/upmpdcli-docker/wiki/Linn-Kazoo-Screenshots)

**Comment by GioF71** [comment]
Hello, version 0.1.16 is available with a few improvements:

- Top Songs now offered as both navigable and list mode
- Multiple artists shown with comma as separator

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.2.0 is available with quite some improvements and new features:

- Initialization changed as suggested by the upmpdcli author (hopefully)
- Manual updated so it reflects the feature list more accurately
- subsonic-connector bumped to newer version (also documented in the manual)
- Genre entry shows Artists and Albums (it was only Albums before)
- Favourite Songs
- Favourite Artists
- Initial View has now group entries for Albums, Artists and Songs so first loading is much faster (less queries)
- Few bugs fixed

These features require subsonic-connector version 0.1.17 so pulling the new upmpdcli image is mandatory.
This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Comment by GioF71** [comment]
Hello, version 0.2.1 is available with quite some improvements and new features:

- Randomness for cover in album presentation
- Fixed album track sorting when there are different artists in the same album

This version is downloaded with the suggested configuration, through the variables:

SUBSONIC_DOWNLOAD_PLUGIN=yes
SUBSONIC_PLUGIN_BRANCH=latest

Remember to keep the docker image updated, so in case the plugin needs new version of some libraries, these will be made available automatically.

**Reply by TobiasDax** [reply]
hi, not sure if there is a better palce to ask questions abou this but i cant seem to get it to work sadly...
When i run the compose with the demo data or my own i get the same problem. The Server shows up in my player (currently tyring with Linn Kazoo and Wiim Home (app))
The server seems to be laoding stuff succsessfully but when i try to access it i get the following output in the logs:

```
CMDTALK: subsonic-app.py: Genre Southern Rock has not art yet, looking for an album
stderr
CMDTALK: subsonic-app.py: Caching genre [Southern Rock] with album_id [f4195a1f22fbb481d01ca08060c42bba]
stderr
CMDTALK: subsonic-app.py: Processing genre [Street Punk]
stderr
CMDTALK: subsonic-app.py: Genre Street Punk has not art yet, looking for an album
stderr
CMDTALK: subsonic-app.py: Caching genre [Street Punk] with album_id [569c478785ca7d6dd8d8927a0930b13a]
stderr
CMDTALK: subsonic-app.py: Processing genre [Tech House]
stderr
CMDTALK: subsonic-app.py: Subsonic Initialization success: [True]
stderr
CMDTALK: subsonic-app.py: Subsonic running
stderr
CMDTALK: subsonic-app.py: browse: args: --{'cmdtalk:proc': 'browse', 'count': '2000', 'offset': '0', 'flag': 'children', 'objid': '0$subsonic$'}--
stderr
CMDTALK: subsonic-app.py: Subsonic Plugin Release 0.2.6
stderr
CMDTALK: subsonic-app.py: browse: path: ----
stderr
CMDTALK: subsonic-app.py: browse: path_list: --['0$subsonic$']-- last: --0$subsonic$--
stderr
CMDTALK: subsonic-app.py: processmessage: processor raised: ['ArtistCover' object has no attribute 'getCoverArt']
stderr
Traceback (most recent call last):
stderr
  File "/usr/share/upmpdcli/cdplugins/pycommon/cmdtalk.py", line 167, in processmessage
stderr
    outfields = processor.process(params)
stderr
                ^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/pycommon/cmdtalkplugin.py", line 64, in process
stderr
    return self.dispatcher.run(params[prcnmkey], params)
stderr
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/pycommon/cmdtalkplugin.py", line 49, in run
stderr
    return func(params)
stderr
           ^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/subsonic-app.py", line 1285, in browse
stderr
    entries = _show_tags(objid, entries)
stderr
              ^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/subsonic-app.py", line 1263, in _show_tags
stderr
    if tag_enabled_in_initial_page(tag): entries.append(tag_to_entry(objid, tag))
stderr
                                                        ^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/subsonic-app.py", line 1254, in tag_to_entry
stderr
    art_id = tag_art_retriever[tagname]()
stderr
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/art_retriever.py", line 54, in group_artists_art_retriever
stderr
    if not art: art = random_artist_art_retriever()
stderr
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/art_retriever.py", line 124, in random_artist_art_retriever
stderr
    return _get_random_artist_cover()
stderr
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/art_retriever.py", line 116, in _get_random_artist_cover
stderr
    return __get_random_artist_cover_by_initial(select_initial)
stderr
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/art_retriever.py", line 108, in __get_random_artist_cover_by_initial
stderr
    return get_artist_cover(artist_id)
stderr
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
stderr
  File "/usr/share/upmpdcli/cdplugins/subsonic/art_retriever.py", line 121, in get_artist_cover
stderr
    return artist_cover.getCoverArt()
stderr
           ^^^^^^^^^^^^^^^^^^^^^^^^
stderr
AttributeError: 'ArtistCover' object has no attribute 'getCoverArt'
stderr
:2:../src/mediaserver/cdplugins/plgwithslave.cxx:585::PlgWithSlave::browse: slave failure
```

**Reply by GioF71** [reply]
Opened issue [here](https://github.com/GioF71/upmpdcli-docker/issues/438) so we don't pollute this repository.

**Comment by danjj84** [comment]
Can see not been many replies in a year, wondering how does this work with Navidrome? Do you need just this plugin then the navidrome library will become UPNP, so can be added as a UPNP source for streamers etc? or is there more too it?

**Reply by GioF71** [reply]
Hello, I have not continued to add posts to this discussion because I thought it had been closed. After all the plugin is not strictly tied to Navidrome and its development cycle. The plugin should work with any subsonic server altough it is actively checked against Lightweight Music Server and Navidrome.
It has evolved in the last year. See the [release notes](https://github.com/GioF71/audio-tools/tree/main/media-servers/subsonic).
Have a look at [this](https://github.com/GioF71/audio-tools/tree/main/media-servers/subsonic) for information on how to setup a docker container and connect upmpdcli to a subsonic server.
Let me know if this helps.

**Comment by deluan** [comment]
Hey @GioF71, wanna create a entry in the new [Apps catalog](https://www.navidrome.org/apps/)?
Just follow the instructions here: https://www.navidrome.org/docs/developers/adding-apps/

I haven't done it myself because we need some kind of screenshot, and a proper project page.

Let me know if you have any questions. Thanks!

**Reply by GioF71** [reply]
Sure, thanks a lot, I will do this ASAP!

**Reply by GioF71** [reply]
Hello, a problem is that this is a plugin for the upmpdcli project. It's not a project on its own.
Is it OK to mention that project?

**Reply by deluan** [reply]
I think the project is the "Subsonic Connector for upnpdcli" or something like this. I don't see a problem with it, as long as it emphasizes the Subsonic (or Navidrome) aspect of it in the description. Like the Sonos bridge (bonob)

**Reply by GioF71** [reply]
Subsonic Connector is just a library, used by the "subsonic plugin for upmpdcli", "mpd-subsonic-scrobbler" and "upnp-scrobbler". The project here might be the "subsonic plugin for upmpdcli" if you agree!

**Reply by deluan** [reply]
Looks good to me
